cirq/contrib/quantum_volume/quantum_volume.py

```diff
@@ -91,45 +92,116 @@ def compute_heavy_set(circuit: cirq.Circuit) -> List[int]:
+        for q in qubits:
+            if q not in mapping:
+                mapping[q] = q
         key = lambda q: mapping[q]
         qubits = frozenset(mapping.keys())
-    circuit_copy = circuit + cirq.measure(*sorted(qubits, key=key))
+
+    # Don't do a single large measurement gate because then the key will be one
+    # large string. Instead, do a bunch of single-qubit measurement gates so we
+    # preserve the qubit keys.
+    sorted_qubits = sorted(qubits, key=key)
+    circuit_copy = circuit + [cirq.measure(q) for q in sorted_qubits]
 
     # Run the sampler to compare each output against the Heavy Set.
-    measurements = sampler.run(program=circuit_copy, repetitions=repetitions)
+    trial_result = sampler.run(program=circuit_copy, repetitions=repetitions)
+
+    # Post-process the results, e.g. to handle error corrections.
+    results = process_results(mapping, compilation_result.parity_map,
+                              trial_result)
 
+    # Aggregate the results into bit-strings (since we are using individual
+    # measurement gates).
+
+    results = results.agg(lambda meas: cirq.value.big_endian_bits_to_int(meas),
+                          axis=1)
     # Compute the number of outputs that are in the heavy set.
-    num_in_heavy_set = np.sum(np.in1d(measurements.data.iloc[:, 0], heavy_set))
+    num_in_heavy_set = np.sum(np.in1d(results, heavy_set))
+
+    # Return the number of Heavy outputs over the number of valid runs.
+    return num_in_heavy_set / len(results)
+
 
-    # Return the number of Heavy outputs over the number of runs.
-    return num_in_heavy_set / repetitions
+def process_results(mapping: Dict[cirq.Qid, cirq.Qid],
+                    parity_mapping: Dict[cirq.Qid, cirq.Qid],
+                    trial_result: cirq.TrialResult) -> pd.DataFrame:
+    """Checks the given results for parity and throws away all of the runs that
+    don't pass the parity test.
+
+    Args:
+        mapping: The circuit's mapping from logical qubit to physical qubit.
+        parity_mapping: The mapping from result qubit to its parity qubit.
+        trial_result: The results to process.
+
+    Returns:
+        Returns the rows that passed the parity test, with the parity qubit
+        measurements removed.
+
+    """
+    bad_measurements = set()
+    # The circuit's mapping from physical qubit to logical qubit.
+    inverse_mapping: Dict[cirq.Qid, cirq.Qid] = {
+        v: k for k, v in mapping.items()
+    }
+
+    for final_qubit, original_qubit in mapping.items():
+        if original_qubit in parity_mapping:
+            parity_qubit = parity_mapping[original_qubit]
+            qubit_meas = trial_result.measurements[str(final_qubit)]
+            final_parity_qubit = inverse_mapping[parity_qubit]
+            parity_meas = trial_result.measurements[str(final_parity_qubit)]
+            # Check each bit's parity qubit results to see if they are correct.
+            for idx, qubit_val in enumerate(qubit_meas):
```

Comment:
I bet this could be faster with vectorized numpy operations

```python
correct_strings = np.ones(dtype=bool)
for dataq, parityq in whatever:
  parity_check = (measurements[dataq] == measurements[parityq])
  correct_strings = np.logical_and(correct_strings, parity_check)
bitstrings = bitstrings[correct_strings]
```

(roughly)

Reply:
Agreed that this should be improved - @Strilanc has a similar idea that uses the Pandas DataFrame directly, so I went with that.